File: DSA/Manav-Rajput/Day32_question2.java
import java.util.List;

class Solution {
    public long findMaxProduct(List<Integer> arr) {
        long MOD = 1000000007;
        if (arr.size() == 1) {
            return arr.get(0) % MOD;
        }
        
        long product = 1;
        int negativeCount = 0;
        int zeroCount = 0;
        int maxNegative = Integer.MIN_VALUE;
        boolean hasPositive = false;
        
        for (int num : arr) {
            if (num == 0) {
                zeroCount++;
                continue;
            }
            
            if (num < 0) {
                negativeCount++;
                maxNegative = Math.max(maxNegative, num);
            } else {
                hasPositive = true;
            }
            
            product = (product * num) % MOD;
        }
        
        if (zeroCount == arr.size() || (negativeCount == 1 && zeroCount + negativeCount == arr.size())) {
            return 0;
        }
        
        if (negativeCount % 2 != 0) {
            product = (product / maxNegative) % MOD;
        }
        
        if (product < 0) {
            product += MOD;
        }
        
        return product;
    }
    
    public static void main(String[] args) {
        Solution solution = new Solution();
        List<Integer> arr1 = List.of(-1, 0, -2, 4, 3);
        List<Integer> arr2 = List.of(-1, 0);
        List<Integer> arr3 = List.of(5);
        
        System.out.println(solution.findMaxProduct(arr1)); // Output: 24
        System.out.println(solution.findMaxProduct(arr2)); // Output: 0
        System.out.println(solution.findMaxProduct(arr3)); // Output: 5
    }
}

File: DSA/Arshdeep-Kaur/Day 32/Day32_Question2.cpp
class Solution {
  public:
    long long int findMaxProduct(vector<int>& arr) {
        if (arr.size() == 1) {
            return arr[0];
        }
        
        long long int pos=1, neg=1, negLar=INT_MIN, mod = 1e9+7, c1=0, c2=0;
        for (int i=0; i<arr.size(); i++) {
            if (arr[i]<0) {
                neg = (neg*(long long int)arr[i])%mod;
                negLar = max((long long int)arr[i], negLar);
                c1++;
            }
            if (arr[i]>0) {
                pos = (pos*(long long int)arr[i])%mod;
                c2++;
            }
        }
        
        if (c1 <= 1 && c2 == 0) {
            return 0;
        }
        if (neg < 0) {
            neg = neg/negLar;
        }
        
        return (pos*neg)%mod;
    }
};
